### DISCUSSION TITLE
[INTERNAL] Performance tests

### DISCUSSION THREAD

**Discussion by Kludex** [Kludex/uvicorn#1472] [Ideas]
### Checklist

- [X] There are no similar issues or pull requests for this yet.
- [ ] I discussed this idea on the [community chat](https://gitter.im/encode/community) and feedback is positive.

### Is your feature related to a problem? Please describe.

`uvicorn` is the fastest ASGI server available for HTTP/1.1, but I'm afraid that time can be dangerous if we don't take preventive measures to maintain its current rank.

The proposal here is nothing exaggerated: just create performance tests. I had this thought when I was looking on the [h11](https://github.com/python-hyper/h11) repository and PRs, as they run their benchmark to make sure that PRs don't drop its performance.

## Describe the solution you would like.

I think we could make use of this example and "amplify" the idea and create an action to be triggered every time we push a commit on a PR, and run this benchmark, receiving a message on that PR saying how long that benchmark took, and compare the results with the latest `master`.  

I'm assuming that the environment is a controlled one, so the same variables will be applied for each execution, and the results are meaningful ones.

## Describe alternatives you considered

1. Don't do anything, it's fine the way it is. Let's be careful on each review.
2. Create this benchmark and ask for collaborators to run it when members are in doubt if the PR could cause a drop in performance.

**Comment by Digitalean** [comment]
Hello,
We have a machine learning use case where using uvicorn with fast api has bad performances.
You can find the results of the test via this link: (https://digitalean.com/performance_python_machine_learning_use_case.pdf)

Can you tell us why we have this performance please? And can you help us to have better performance with uvicorn and fast api please?

Thanks in advance.

**Comment by Kludex** [comment]
cc @encode/maintainers

**Comment by euri10** [comment]
I have nothing against that
